kolibri/auth/models.py

```diff
@@ -1,497 +1,428 @@
+    # https://django-mptt.github.io/django-mptt/models.html#treeforeignkey-treeonetoonefield-treemanytomanyfield
+    collection = TreeForeignKey("Collection")
+    kind = models.CharField(max_length=20, choices=KINDS)
 
-class FacilityAdminManager(models.Manager):
-    def get_queryset(self):
-        return super(FacilityAdminManager, self).get_queryset().filter(kind='FacilityAdmin')
+    class Meta:
+        unique_together = (("user", "collection", "kind"),)
 
+    def infer_dataset(self):
+        user_dataset = self.user.dataset
+        collection_dataset = self.collection.dataset
+        if user_dataset != collection_dataset:
+            raise KolibriValidationError("The collection and user for a Role object must be in the same dataset.")
+        return user_dataset
 
-class CoachManager(models.Manager):
-    def get_queryset(self):
-        return super(CoachManager, self).get_queryset().filter(kind='Coach')
 
+class FacilityManager(models.Manager):
 
-class LearnerManager(models.Manager):
     def get_queryset(self):
-        return super(LearnerManager, self).get_queryset().filter(kind='Learner')
-
+        return super(FacilityManager, self).get_queryset().filter(kind=Collection.KIND_FACILITY)
 
-class FacilityAdmin(Role):
-    objects = FacilityAdminManager()
-
-    class Meta:
-        proxy = True
-
-    def save(self, *args, **kwargs):
-        self.kind = "FacilityAdmin"
-        return super(FacilityAdmin, self).save(*args, **kwargs)
 
+class Facility(Collection):
 
-class Coach(Role):
-    objects = CoachManager()
+    objects = FacilityManager()
 
     class Meta:
         proxy = True
 
     def save(self, *args, **kwargs):
-        self.kind = "Coach"
-        return super(Coach, self).save(*args, **kwargs)
+        if self.parent:
+            raise IntegrityError("Facility must be the root of a collection tree, and cannot have a parent.")
+        self.kind = Collection.KIND_FACILITY
+        super(Facility, self).save(*args, **kwargs)
+
+    def infer_dataset(self):
+        # if we don't yet have a dataset, create a new one for this facility
+        if not self.dataset_id:
+            self.dataset = FacilityDataset.objects.create()
+        return self.dataset
+
+    def get_classrooms(self):
+        """
+        Returns a QuerySet of Classrooms under this Facility.
 
+        :return: A Classroom QuerySet.
+        """
+        return Classroom.objects.filter(parent=self)
 
-class Learner(Role):
-    objects = LearnerManager()
+    def add_admin(self, user):
+        return self.add_user(user, Role.KIND_ADMIN)
 
-    class Meta:
-        proxy = True
+    def add_admins(self, users):
+        return [self.add_admin(user) for user in users]
 
-    def save(self, *args, **kwargs):
-        self.kind = "Learner"
-        return super(Learner, self).save(*args, **kwargs)
+    def remove_admin(self, user):
+        self.remove_user(user, Role.KIND_ADMIN)
 
+    def add_coach(self, user):
```

Comment:
I recall we talked about the idea. I didn't consider it settled (sorry if I didn't communicate that well in my above comment).

Somehow whether the roles coach->facility and admin->classroom exist is a matter of taste. The role coach->facility is equivalent to being a coach for *all* classrooms in a facility, and I can't think of a permission which should distinguish admin->classroom and coach->classroom (though I'll take you at your word if you say there is). For reference, see all the permissions I *could* think of: http://kolibri.readthedocs.org/en/latest/dev/uap/permissions.html#data-access-permissions

That's my opinion, however I find that we're now discussing ~6 lines of code to death, so I leave it to you.